### PR TITLE
First commit of cobra and viper usage.

### PR DESCRIPTION
**Pull request by timorunge**
**WIP**

*This is not meant to be merged yet.*

@gammons let me know what you think and how you're adding dependencies for vendors. To be honest I just tried `govendor`. :)

# new command structure

## add

`ultralist (a|add) [todo]` -> `ulatrlist (a|add) [todo]`
`ultralist done [todo]` -> `ultralist (a|add) [todo] --done`

## list

`ultralist (l|list)` -> `ultralist (l|list)`

`ultralist ln` -> `ultralist (l|list) (n|notes) [todo_id]`

`ultralist agenda` -> `ultralist agenda`

`ultralist (l|list) overdue` -> `ultralist (l|list) overdue`

### new

`ultralist (l|list) archive`

## delete

`ultralist (d|delete) [todo_id]` -> `ultralist (d|delete) [todo_id]`

## complete

`ultralist (c|complete) [todo_id]` -> `ultralist (c|complete) [todo_id]`
`ultralist (uc|uncomplete) [todo_id]` -> `ultralist (c|complete) [todo_id]
--revert`

## archive

`ultralist ac` -> `ultralist (ar|archive) (c|completed)`
`ultralist (ar|archive) [todo_id]` -> `ultralist (ar|archive) [todo_id]`
`ultralist (uar|unarchive) [todo_id]` -> `ultralist (ar|archive) [todo_id]
--revert`

### New

`ultralist (ar|archive) (gc|garbage-collect)` replaces `ultralist gc`
`ultralist (ar|archive) (l|list)`

## edit

`ultralist (e|edit) [todo_id]` -> `ultralist (e|edit) [todo_id]`

## expand

`ultralist (ex|expand) [???]` -> `ultralist (ex|expand) [project]: [todo_1], [todo_2], [todo_3], ...`

## notes

`ultralist an [todo_id]` -> `ultralist (a|add) (n|note) [todo_id]`

`ultralist n [todo_id]` -> `ultralist (l|list) (n|notes) [todo_id]`

`ultralist dn [todo_id] [note_id]` -> `ultralist (d|delete) (n|note) [todo_id]
[note_id]`

`ultralist en [todo_id] [note_id]` -> `ultralist (e|edit) (n|note) [todo_id] [note_id]`

## prioritize

`ultralist (p|prioritize) [todo_id]` -> `ultralist (p|prioritize) [todo_id]`

`ultralist (up|unprioritize) [todo_id]` -> `ultralist (p|prioritize) [todo_id]
--revert`

## init

`ultralist init` -> `ultralist init`

## web

`ultralist web` -> `ultralist web`

### sync

`ultralist sync` -> `ultralist web sync`

### auth

`ultralist auth` -> `ultralist web auth`

### check_auth

`ultralist check_auth` -> `ultralist web auth check`

## new commands

`ultralist version `

`ultralist help`

### CODE REVIEW

File: ultralist/app.go [hunk 1]
```diff
@@ -236,7 +236,6 @@ func (a *App) ExpandTodo(input string) {
 		a.AddTodo(strings.Join(args, ""))
 	}
 
-	a.TodoList.Delete(id)
```
Comment (timorunge) [inline ultralist/app.go]: I don't understand why this was here. It was always deleting one of the new todos.

File: ultralist/date_filter.go [hunk 2]
```diff
@@ -188,9 +188,11 @@ func (f *DateFilter) filterOverdue(pivot time.Time) []*Todo {
 	pivotDate := pivot.Format("2006-01-02")
 
 	for _, todo := range f.Todos {
-		dueTime, _ := time.ParseInLocation("2006-01-02", todo.Due, f.Location)
-		if dueTime.Before(pivot) && pivotDate != todo.Due {
-			ret = append(ret, todo)
+		if todo.Due != "" {
```
Comment (timorunge) [inline ultralist/date_filter.go]: The overdue filter was always displaying todos without due date.

File: ultralist/filter.go [hunk 3]
```diff
@@ -35,22 +36,26 @@ func (f *TodoFilter) isFilteringByContexts(input string) bool {
 
 func (f *TodoFilter) filterArchived(input string) []*Todo {
 
-	// do not filter archived if want completed items
-	completedRegex, _ := regexp.Compile(`completed`)
```
Comment (timorunge) [inline ultralist/filter.go]: Separation makes it easier.

File: cmd/root.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+// Execute adds all child commands to the root command and sets flags appropriately.
+func Execute() {
+	if err := rootCmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func init() {
+	// cobra.OnInitialize(initConfig)
```
Comment (timorunge) [inline cmd/root.go]: Not required right now. Kept it for the case if we'd like to make ultralist configurable (e.g. colors).

File: cmd/list.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+  Lists all todos due today for +project, grouped by context.
+
+  ulralist list @frank due tom by p
+  Lists all todos due tomorrow concerining @frank for +project, grouped by project.`
+	listCmdLongDesc = listCmdDesc
+)
+
+var listCmd = &cobra.Command{
+	Use:     "list",
+	Aliases: []string{"l", "ls"},
```
Comment (timorunge) [inline cmd/list.go]: `ls` for more unix style. :)

File: cmd/delete.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+var (
+	deleteCmdDesc    = "Deletes todos and notes"
+	deleteCmdExample = `  ultralist delete 33
+  Deletes todo with id 33.`
+	deleteCmdLongDesc = deleteCmdDesc
+)
+
+var deleteCmd = &cobra.Command{
+	Use:     "delete [id]",
+	Aliases: []string{"d", "rm"},
```
Comment (timorunge) [inline cmd/delete.go]: `rm` for more unix style. :)

Comment (gammons) [reply cmd/delete.go]: love it.

File: cmd/list.go (hunk 5 above)
Comment (gammons) [reply cmd/list.go]: also love it

File: cmd/root.go (hunk 4 above)
Comment (gammons) [reply cmd/root.go]: that will be the next step.  the [color](https://github.com/fatih/color) package has a `NoColor` flag that we can hook into, and will solve a lot of problems for windows users.

File: ultralist/app.go (hunk 1 above)
Comment (gammons) [reply ultralist/app.go]: The original intention of expanding a todo was to create many todos out of a single one.  I think we should talk about this change in a separate PR.

File: ultralist/date_filter.go (hunk 2 above)
Comment (gammons) [reply ultralist/date_filter.go]: could you push this into a separate PR?  I agree that todos without a due date shouldn't be part of the `overdue` filter.

File: cmd/root.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+
+	homedir "github.com/mitchellh/go-homedir"
```
Comment (gammons) [inline cmd/root.go]: you can get the user's homedir using internal go libraries, kind of like I did here:

https://github.com/ultralist/ultralist/blob/master/ultralist/file_store.go#L85

**Comment by gammons** [comment]
ultralist uses go modules now, that has reasonable documentation.  read `go helo modules `.  But the TLDR is essentially to run `go mod vendor` to vendor up the dependencies.

File: ultralist/date_filter.go (hunk 2 above)
Comment (timorunge) [reply ultralist/date_filter.go]: See #169 and #170.

File: cmd/root.go (hunk 7 above)
Comment (timorunge) [reply cmd/root.go]: That's what comes with cobra by default. Reason is described in https://github.com/mitchellh/go-homedir: 

*Why not just use os/user? The built-in os/user package requires cgo on Darwin systems. This means that any Go code that uses that package cannot cross compile. But 99% of the time the use for os/user is just to retrieve the home directory, which we can do for the current user without cgo. This library does that, enabling cross-compilation.*

File: ultralist/app.go (hunk 1 above)
Comment (timorunge) [reply ultralist/app.go]: Removed.

File: cmd/root.go (hunk 7 above)
Comment (gammons) [reply cmd/root.go]: @timorunge hmm, but it looks like that library might be out of date.  Take a look at this: https://github.com/mitchellh/go-homedir/issues/20

Comment (timorunge) [reply cmd/root.go]: Yeah, it's also pointing [`os.UserHomeDir()`](https://golang.org/pkg/os/#UserHomeDir) which is even more simplistic.

**Comment by timorunge** [comment]
@TODO: Add `ultralist init config` to init `~/.ultralist.yaml` or `~/.config/ultralist/config.yaml` (I prefer this one since credentials are stored already in this folder).

@gammons configuration can be done either in `yaml`, `json`, `toml` or `hcl`. I personally prefer `yaml` but I'd like to keep this up to you.

**Comment by gammons** [comment]
@timorunge what do we need a config file for?

**Comment by timorunge** [comment]
@gammons e.g. for enabling and disabling colours. At the same time there might be some way to define own colours and / or other ways of modifying the application (e.g. in the long run define the way how ultralist looks, moving the check boxes to the right, define an own date format etc.).

**Comment by gammons** [comment]
I'd like to control that with a `--nocolor` flag and stay away from a config file as long as I can.

**Comment by timorunge** [comment]
@gammons sure. Having a configuration files does not mean that we have to use it.

I'll skip the config part for the time being even if I'd prefer some custom styling. :)

**Comment by gammons** [comment]
one other thing, you'll need to run `go mod vendor` as well.

**Comment by timorunge** [comment]
@gammons Yeah, I'm fully aware. I need to do some final testing and want to review everything once again. :)

**Comment by gammons** [comment]
I think overall this PR looks great.  I really appreciate your work here.  I'm going to release this as ultralist 1.0!

**Comment by gammons** [comment]
this looks good now.  how do you feel about me merging it?

**Comment by timorunge** [comment]
@gammons Let's do it. But please squash merge (we should generally speaking do the same for PRs) it to that we can revert easier.

**Comment by gammons** [comment]
Thanks for your work on this @timorunge !  I'm going to run it for a couple of days and then I'll cut release v1.0.

**Comment by timorunge** [comment]
Sure, @gammons. I have some more ideas, I'll start to work on them as soon as I'll have some time and submit some PRs. :)

**Pull request merged**